File: 240202/2차원 최대 증가 수열/longest-increasing-sequence-2d.java
import java.util.Scanner;

public class Main {

    private static int n, m;
    private static int[][] board, dp;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        n = sc.nextInt();
        m = sc.nextInt();


        board = new int[n][m];
        dp = new int[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                board[i][j] = sc.nextInt();
                dp[i][j] = -1;
            }
        }

        int maxAnswer = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                maxAnswer = Math.max(dpf(i, j), maxAnswer);
            }
        }

        System.out.print(maxAnswer);

        // for (int i = 0; i < n; i++) {
        //     for (int j = 0; j < m; j++) {
        //         System.out.print(dp[i][j] + " ");
        //     }
        //     System.out.println();
        // }


    } 
    // dpf(r, c) = (0, 0) ~ (r, c) 도착할 때까지의 최대 칸의 수
    // dpf(r, c) = max(dpf(preR, preC) + 1, ret) where 0 <= preR < R and 0 <= preC < c 
    // and board[preR][preC] < board[r][c]
    private static int dpf(int r, int c) {
        if (dp[r][c] != -1) {
            return dp[r][c];
        }
        // 최소는 1이다.
        int ret = 0;
        for (int preR = 0; preR < r; preR++) {
            for (int preC = 0; preC < c; preC++) {
                if (board[preR][preC] < board[r][c]){
                    ret = Math.max(dpf(preR, preC) + 1, ret);
                }
            }
        }


        if (r == 0 && c == 0) {
            return 1;
        }


        dp[r][c] = ret;
        return ret;
    }
}

File: 240217/2차원 최대 증가 수열/longest-increasing-sequence-2d.py
n, m = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(n)]
dp = [[-1 for _ in range(m)] for _ in range(n)]

inf = int(1e9)
def dpf(r, c):
    if dp[r][c] != -1:
        return dp[r][c]

    if r == 0 and c == 0:
        return 1

    ret = -inf
    for i in range(0, r):
        for j in range(0, c):
            if board[r][c] > board[i][j]:
                ret = max(ret, dpf(i, j) + 1)

    dp[r][c] = ret
    return ret

ans = 0
for i in range(n):
    for j in range(m):
        ans = max(ans, dpf(i, j))

print(ans)